Source Code Pro font can not be installed under JDK11

When i run Eclipse 2018-09 with JDK11 the Source Code Pro font cannot be installed because of using sun.misc.Unsafe internally.

java.lang.ClassNotFoundException: An error occurred while automatically activating bundle com.codeaffine.eclipse.ui.swt.theme (23).
	at org.eclipse.osgi.internal.hooks.EclipseLazyStarter.postFindLocalClass(EclipseLazyStarter.java:126)
	at org.eclipse.osgi.internal.loader.classpath.ClasspathManager.findLocalClass(ClasspathManager.java:557)
	at org.eclipse.osgi.internal.loader.ModuleClassLoader.findLocalClass(ModuleClassLoader.java:331)
	at org.eclipse.osgi.internal.loader.BundleLoader.findLocalClass(BundleLoader.java:395)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClassInternal(BundleLoader.java:473)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:422)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:414)
	at org.eclipse.osgi.internal.loader.ModuleClassLoader.loadClass(ModuleClassLoader.java:153)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:521)
	at org.eclipse.osgi.internal.framework.EquinoxBundle.loadClass(EquinoxBundle.java:612)
	at org.eclipse.core.internal.registry.osgi.RegistryStrategyOSGI.createExecutableExtension(RegistryStrategyOSGI.java:180)
	at org.eclipse.core.internal.registry.ExtensionRegistry.createExecutableExtension(ExtensionRegistry.java:934)
	at org.eclipse.core.internal.registry.ConfigurationElement.createExecutableExtension(ConfigurationElement.java:246)
	at org.eclipse.core.internal.registry.ConfigurationElementHandle.createExecutableExtension(ConfigurationElementHandle.java:63)
	at org.eclipse.e4.ui.css.core.impl.engine.RegistryCSSPropertyHandlerProvider.registerPropertyHandler(RegistryCSSPropertyHandlerProvider.java:115)
	at org.eclipse.e4.ui.css.core.impl.engine.RegistryCSSPropertyHandlerProvider.configure(RegistryCSSPropertyHandlerProvider.java:83)
	at org.eclipse.e4.ui.css.core.impl.engine.RegistryCSSPropertyHandlerProvider.<init>(RegistryCSSPropertyHandlerProvider.java:66)
	at org.eclipse.e4.ui.css.swt.engine.CSSSWTEngineImpl.initializeCSSPropertyHandlers(CSSSWTEngineImpl.java:67)
	at org.eclipse.e4.ui.css.swt.engine.AbstractCSSSWTEngineImpl.<init>(AbstractCSSSWTEngineImpl.java:72)
	at org.eclipse.e4.ui.css.swt.engine.CSSSWTEngineImpl.<init>(CSSSWTEngineImpl.java:43)
	at org.eclipse.e4.ui.css.swt.internal.theme.ThemeEngineManager.getCSSSWTEngine(ThemeEngineManager.java:52)
	at org.eclipse.e4.ui.css.swt.internal.theme.ThemeEngineManager.getEngineForDisplay(ThemeEngineManager.java:41)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.createThemeEngine(PartRenderingEngine.java:1432)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.initializeStyling(PartRenderingEngine.java:1309)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1066)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:339)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1062)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:156)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:628)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:339)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:563)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:151)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:155)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:199)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:137)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:107)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:391)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:246)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:659)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:595)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1501)
Caused by: org.osgi.framework.BundleException: Error starting module.
	at org.eclipse.osgi.container.Module.doStart(Module.java:593)
	at org.eclipse.osgi.container.Module.start(Module.java:452)
	at org.eclipse.osgi.framework.util.SecureAction.start(SecureAction.java:471)
	at org.eclipse.osgi.internal.hooks.EclipseLazyStarter.postFindLocalClass(EclipseLazyStarter.java:117)
	... 44 more
Caused by: java.lang.IllegalStateException: java.lang.NoSuchMethodException: sun.misc.Unsafe.defineClass(java.lang.String, [B, int, int, java.lang.ClassLoader, java.security.ProtectionDomain)
	at com.codeaffine.eclipse.swt.util.Unsafe.<init>(Unsafe.java:30)
	at com.codeaffine.eclipse.swt.util.ControlReflectionUtil.<init>(ControlReflectionUtil.java:50)
	at com.codeaffine.eclipse.swt.widget.scrollable.ScrollableAdapterFactory.<init>(ScrollableAdapterFactory.java:51)
	at com.codeaffine.eclipse.ui.swt.theme.ScrollbarPreferenceApplicator.<init>(ScrollbarPreferenceApplicator.java:48)
	at com.codeaffine.eclipse.ui.swt.theme.Activator.<init>(Activator.java:23)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at java.base/java.lang.Class.newInstance(Class.java:584)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.loadBundleActivator(BundleContextImpl.java:766)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.start(BundleContextImpl.java:719)
	at org.eclipse.osgi.internal.framework.EquinoxBundle.startWorker0(EquinoxBundle.java:1005)
	at org.eclipse.osgi.internal.framework.EquinoxBundle$EquinoxModule.startWorker(EquinoxBundle.java:357)
	at org.eclipse.osgi.container.Module.doStart(Module.java:584)
	... 47 more
Caused by: java.lang.NoSuchMethodException: sun.misc.Unsafe.defineClass(java.lang.String, [B, int, int, java.lang.ClassLoader, java.security.ProtectionDomain)
	at java.base/java.lang.Class.getMethod(Class.java:2109)
	at com.codeaffine.eclipse.swt.util.Unsafe.getDefineClassMethod(Unsafe.java:65)
	at com.codeaffine.eclipse.swt.util.Unsafe.<init>(Unsafe.java:28)
	... 61 more

--- a/com.codeaffine.eclipse.swt.test/src/com/codeaffine/eclipse/swt/util/UnsafeTest.java
+++ b/com.codeaffine.eclipse.swt.test/src/com/codeaffine/eclipse/swt/util/UnsafeTest.java
@@ -14,7 +14,6 @@ import static com.codeaffine.eclipse.swt.testhelper.TestResources.CLASS_NAME;
 import static com.codeaffine.eclipse.swt.testhelper.TestResources.LOCATION;
 import static com.codeaffine.eclipse.swt.testhelper.TestResources.PROTECTED_CLASS_NAME;
 import static com.codeaffine.eclipse.swt.testhelper.TestResources.PROTECTED_LOCATION;
-import static com.codeaffine.test.util.lang.ThrowableCaptor.thrownBy;
 import static org.assertj.core.api.Assertions.assertThat;
 
 import java.security.ProtectionDomain;
@@ -38,7 +37,7 @@ public class UnsafeTest {
   public void defineClass() {
     byte[] bytes = resourceLoader.load( LOCATION );
 
-    Class<?> actual = unsafe.defineClass( CLASS_NAME, bytes, 0, bytes.length, getLoader(), getDomain() );
+    Class<?> actual = unsafe.defineClass( CLASS_NAME, bytes, 0, bytes.length, getLoader(), getDomain(), getClass() );
 
     assertThat( actual.getName() ).isEqualTo( CLASS_NAME );
   }
@@ -48,7 +47,7 @@ public class UnsafeTest {
     ensureSignerInfoInCurrentClassLoader();
     byte[] bytes = resourceLoader.load( PROTECTED_LOCATION );
 
-    Class<?> actual = unsafe.defineClass( PROTECTED_CLASS_NAME, bytes, 0, bytes.length, getLoader(), getDomain() );
+    Class<?> actual = unsafe.defineClass( PROTECTED_CLASS_NAME, bytes, 0, bytes.length, getLoader(), getDomain(), Widget.class );
 
     assertThat( actual.getName() ).isEqualTo( PROTECTED_CLASS_NAME );
   }
@@ -57,20 +56,10 @@ public class UnsafeTest {
   public void defineClassTwice() {
     byte[] bytes = resourceLoader.load( LOCATION );
 
-    Class<?> first = unsafe.defineClass( CLASS_NAME, bytes, 0, bytes.length, getLoader(), getDomain() );
-    Class<?> second = unsafe.defineClass( CLASS_NAME, bytes, 0, bytes.length, getLoader(), getDomain() );
+    Class<?> first = unsafe.defineClass( CLASS_NAME, bytes, 0, bytes.length, getLoader(), getDomain(), getClass() );
+    Class<?> second = unsafe.defineClass( CLASS_NAME, bytes, 0, bytes.length, getLoader(), getDomain(), getClass() );
 
     assertThat( first ).isSameAs( second );
-  }
-
-  @Test
-  public void defineClassThatDoesNotExist() {
-    byte[] bytes = resourceLoader.load( LOCATION );
-
-    Throwable actual = thrownBy(
-      () ->  unsafe.defineClass( "unknown", bytes, 0, bytes.length, getLoader(), getDomain() ) );
-
-    assertThat( actual ).hasRootCauseInstanceOf( NoClassDefFoundError.class );
   }
 
   @Test

--- a/com.codeaffine.eclipse.swt/src/com/codeaffine/eclipse/swt/util/ControlReflectionUtil.java
+++ b/com.codeaffine.eclipse.swt/src/com/codeaffine/eclipse/swt/util/ControlReflectionUtil.java
@@ -51,7 +51,7 @@ public class ControlReflectionUtil {
   }
 
   public Class<? extends Widget> defineWidgetClass( String name ) {
-    return execute( () -> defineClass( name ) );
+    return execute( () -> defineClass( name, Widget.class ) );
   }
 
   public <T extends Widget> T newInstance( Class<T> type ) {
@@ -75,12 +75,12 @@ public class ControlReflectionUtil {
   }
 
   @SuppressWarnings("unchecked")
-  private Class<? extends Widget> defineClass( String name ) {
+  private Class<? extends Widget> defineClass( String name, Class<?> otherInPackage ) {
     String path = name.replaceAll( "\\.", "/" ) + ".class";
     byte[] bytes = new ResourceLoader().load( path );
     ClassLoader loader = getClass().getClassLoader();
     ProtectionDomain domain = getClass().getProtectionDomain();
-    return ( Class<? extends Widget> )unsafe.defineClass( name, bytes, 0, bytes.length, loader, domain  );
+    return ( Class<? extends Widget> )unsafe.defineClass( name, bytes, 0, bytes.length, loader, domain, otherInPackage );
   }
 
   private <T extends Widget> T createNewIewInstance( Class<T> type ) throws Exception {

--- a/com.codeaffine.eclipse.swt/src/com/codeaffine/eclipse/swt/util/Unsafe.java
+++ b/com.codeaffine.eclipse.swt/src/com/codeaffine/eclipse/swt/util/Unsafe.java
@@ -10,6 +10,8 @@
  */
 package com.codeaffine.eclipse.swt.util;
 
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodHandles.Lookup;
 import java.lang.reflect.Field;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
@@ -20,12 +22,20 @@ public class Unsafe {
   private Object unsafeInstance;
   private Method allocateInstance;
   private Method defineClass;
+  private Method privateLookupIn;
 
   public Unsafe() {
     try {
       unsafeInstance = getUnsafeInstance();
       allocateInstance = getAllocateInstanceMethod();
-      defineClass = getDefineClassMethod();
+      try {
+        defineClass = getUnsafeDefineClassMethod();
+      } catch( NoSuchMethodException noSuchMethod ) {
+        // These methods are part of JDK9+'s public API. However, getting them
+        // via reflection ensures that the code still compiles on JDK8.
+        defineClass = getLookupDefineClassMethod();
+        privateLookupIn = getPrivateLookupInMethod();
+      }
     } catch( Exception unresolvable ) {
       throw new IllegalStateException( unresolvable );
     }
@@ -42,11 +52,11 @@ public class Unsafe {
   }
 
   public Class<?> defineClass(
-    String name, byte[] bytes, int offset, int length, ClassLoader loader, ProtectionDomain domain )
+    String name, byte[] bytes, int offset, int length, ClassLoader loader, ProtectionDomain domain, Class<?> otherInPackage )
   {
     Class<?> result = loadFromClassLoader( name, loader );
     if( result == null ) {
-      result = defineClassUnsafe( name, bytes, offset, length, loader, domain );
+      result = ( Class<?> )invokeDefineClass( name, bytes, offset, length, loader, domain, otherInPackage );
     }
     return result;
   }
@@ -61,13 +71,21 @@ public class Unsafe {
     return getUnsafeClass().getMethod( "allocateInstance", Class.class );
   }
 
-  private static Method getDefineClassMethod() throws Exception {
+  private static Method getUnsafeDefineClassMethod() throws Exception {
     return getUnsafeClass().getMethod(
       "defineClass", String.class, byte[].class, int.class, int.class, ClassLoader.class, ProtectionDomain.class );
   }
 
   private static Class<?> getUnsafeClass() throws ClassNotFoundException {
     return Unsafe.class.getClassLoader().loadClass( "sun.misc.Unsafe" );
+  }
+
+  private static Method getLookupDefineClassMethod() throws Exception {
+    return Lookup.class.getMethod( "defineClass", byte[].class );
+  }
+
+  private static Method getPrivateLookupInMethod() throws Exception {
+    return MethodHandles.class.getMethod( "privateLookupIn", Class.class, Lookup.class );
   }
 
   private static Class<?> loadFromClassLoader( String name, ClassLoader loader ) {
@@ -80,12 +98,14 @@ public class Unsafe {
     }
   }
 
-  private Class<?> defineClassUnsafe(
-    String name, byte[] bytes, int offset, int length, ClassLoader loader, ProtectionDomain domain )
+  private Object invokeDefineClass(
+    String name, byte[] bytes, int offset, int length, ClassLoader loader, ProtectionDomain domain, Class<?> otherInPackage )
   {
     try {
-      return ( Class<?> )defineClass.invoke(
-        unsafeInstance, name, bytes, Integer.valueOf( offset), Integer.valueOf( length ), loader, domain );
+      if (privateLookupIn != null) {
+        return defineClass.invoke( privateLookupIn.invoke( null, otherInPackage, MethodHandles.lookup() ), bytes );
+      }
+      return defineClass.invoke( unsafeInstance, name, bytes, Integer.valueOf( offset ), Integer.valueOf( length ), loader, domain );
     } catch( RuntimeException rte ) {
       throw rte;
     } catch( InvocationTargetException ite ) {

